node/src/exchange/client.rs

```diff
@@ -56,63 +96,697 @@ impl ExchangeClientHandler {
+                if headers[0].hash().as_bytes() != hash {
+                    return Err(ExchangeError::InvalidResponse);
                 }
             }
+
+            _ => return Err(ExchangeError::InvalidResponse),
         }
 
-        respond_to.maybe_send_ok(headers);
+        Ok(headers)
     }
 
     #[instrument(level = "trace", skip(self))]
-    pub(super) fn on_failure(&mut self, request_id: RequestId, error: OutboundFailure) {
+    pub(super) fn on_failure(&mut self, peer: PeerId, request_id: R::Id, error: OutboundFailure) {
         trace!("Outbound failure");
 
-        if let Some(req_info) = self.reqs.remove(&request_id) {
-            req_info
+        if let Some(state) = self.reqs.remove(&request_id) {
+            state
                 .respond_to
                 .maybe_send_err(ExchangeError::OutboundFailure(error));
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::exchange::utils::ExtendedHeaderExt;
+    use celestia_proto::p2p::pb::header_request::Data;
+    use celestia_proto::p2p::pb::StatusCode;
+    use celestia_types::consts::HASH_SIZE;
+    use celestia_types::{DataAvailabilityHeader, Hash, ValidatorSet};
+    use std::collections::VecDeque;
+    use std::sync::atomic::{AtomicU64, Ordering};
+    use tendermint::block::header::Version;
+    use tendermint::block::{Commit, Header, Id};
+    use tendermint::hash::AppHash;
+    use tendermint::Time;
+
+    #[tokio::test]
+    async fn request_heigh() {
+        let peer_tracker = peer_tracker_with_n_peers(15);
+        let mut mock_req = MockReq::new();
+        let mut handler = ExchangeClientHandler::<MockReq>::new(peer_tracker);
+
+        let (tx, rx) = oneshot::channel();
+
+        handler.on_send_request(&mut mock_req, HeaderRequest::with_origin(5, 1), tx);
+
+        let expected = gen_header_response(5);
+        let expected_header = expected.to_extended_header().unwrap();
+
+        mock_req.send_n_responses(&mut handler, 1, vec![expected]);
+
+        let result = rx.await.unwrap().unwrap();
+        assert_eq!(result.len(), 1);
+        assert_eq!(result[0], expected_header);
+    }
+
+    #[tokio::test]
+    async fn request_hash() {
+        let peer_tracker = peer_tracker_with_n_peers(15);
+        let mut mock_req = MockReq::new();
+        let mut handler = ExchangeClientHandler::<MockReq>::new(peer_tracker);
+
+        let (tx, rx) = oneshot::channel();
+        let expected = gen_header_response(5);
+        let expected_header = expected.to_extended_header().unwrap();
+
+        handler.on_send_request(
+            &mut mock_req,
+            HeaderRequest::with_hash(expected_header.hash()),
+            tx,
+        );
+
+        mock_req.send_n_responses(&mut handler, 1, vec![expected]);
+
+        let result = rx.await.unwrap().unwrap();
+        assert_eq!(result.len(), 1);
+        assert_eq!(result[0], expected_header);
+    }
+
+    #[tokio::test]
+    async fn request_range() {
```

Comment:
It'd be nice to also have a test that range requests sort the returned headers by height, or is it implementation detail?

Reply:
The current Celestia implementation returns the results sorted. But I decided to not trust it. I can add a test if you want.

Reply:
yeah it's probably because `Store` expects it to be sorted. I think the better place for that check would be in the `Store` itself. It can mean double sorting, but it is fine if we stick to the default `pdqsort` as it's O(n) for ascending / descending sequences

Reply:
added a test for this